### ISSUE TITLE
Cannot authenticate with SoundCloud API anymore

### ISSUE THREAD

**Issue by matthewmcvickar** [soundcloud/api#295]
## Cannot authenticate with API

I am consistently unable to authenticate with the SoundCloud API starting on June 6, 2024. On that day, at first, I was able to authenticate intermittently, but for the last three days I have been getting the same response repeatedly.

Up to this point, I had been authenticating successfully on a regular basis.

I have updated my code to use the new OAuth 2.1 authentication method, and am still getting a 401 error when trying to authenticate.

I have double-checked my client ID and client secret, and neither of them were changed.

I am additionally receiving the error message `invalid_client`.

### Endpoint(s):
* `POST https://secure.soundcloud.com/oauth/token`

### Scope(s):
* None (application is not using authentication i.e., implicit flow)

### Steps to reproduce:

```sh
curl -v -X POST "https://secure.soundcloud.com/oauth/token" \
     -H  "accept: application/json; charset=utf-8" \
     -H  "Content-Type: application/x-www-form-urlencoded" \
     -H  "Authorization: Basic Base64(XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX:XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX)" \
     --data-urlencode "grant_type=client_credentials"
```

### Expected behaviour:

I expect to receive an auth token.

### Actual behaviour:

I consistently receive a 401 error. Here is the entirety of the request and the response (with my client ID and secret XXX'd out).

```
curl -v -X POST "https://secure.soundcloud.com/oauth/token" \
     -H  "accept: application/json; charset=utf-8" \
     -H  "Content-Type: application/x-www-form-urlencoded" \
     -H  "Authorization: Basic Base64(XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX:XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX)" \
     --data-urlencode "grant_type=client_credentials"

Note: Unnecessary use of -X or --request, POST is already inferred.
* Host secure.soundcloud.com:443 was resolved.
* IPv6: (none)
* IPv4: 52.84.162.44, 52.84.162.41, 52.84.162.43, 52.84.162.123
*   Trying 52.84.162.44:443...
* Connected to secure.soundcloud.com (52.84.162.44) port 443
* ALPN: curl offers h2,http/1.1
* (304) (OUT), TLS handshake, Client hello (1):
*  CAfile: /etc/ssl/cert.pem
*  CApath: none
* (304) (IN), TLS handshake, Server hello (2):
* (304) (IN), TLS handshake, Unknown (8):
* (304) (IN), TLS handshake, Certificate (11):
* (304) (IN), TLS handshake, CERT verify (15):
* (304) (IN), TLS handshake, Finished (20):
* (304) (OUT), TLS handshake, Finished (20):
* SSL connection using TLSv1.3 / AEAD-AES128-GCM-SHA256 / [blank] / UNDEF
* ALPN: server accepted h2
* Server certificate:
*  subject: CN=*.soundcloud.com
*  start date: Feb  6 12:22:15 2024 GMT
*  expire date: Mar  9 12:22:14 2025 GMT
*  subjectAltName: host "secure.soundcloud.com" matched cert's "*.soundcloud.com"
*  issuer: C=BE; O=GlobalSign nv-sa; CN=GlobalSign GCC R3 DV TLS CA 2020
*  SSL certificate verify ok.
* using HTTP/2
* [HTTP/2] [1] OPENED stream for https://secure.soundcloud.com/oauth/token
* [HTTP/2] [1] [:method: POST]
* [HTTP/2] [1] [:scheme: https]
* [HTTP/2] [1] [:authority: secure.soundcloud.com]
* [HTTP/2] [1] [:path: /oauth/token]
* [HTTP/2] [1] [user-agent: curl/8.6.0]
* [HTTP/2] [1] [accept: application/json; charset=utf-8]
* [HTTP/2] [1] [content-type: application/x-www-form-urlencoded]
* [HTTP/2] [1] [authorization: Basic Base64(XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX:XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX)]
* [HTTP/2] [1] [content-length: 29]
> POST /oauth/token HTTP/2
> Host: secure.soundcloud.com
> User-Agent: curl/8.6.0
> accept: application/json; charset=utf-8
> Content-Type: application/x-www-form-urlencoded
> Authorization: Basic Base64(XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX:XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX)
> Content-Length: 29
>
< HTTP/2 401
< content-type: application/json; charset=utf-8
< content-length: 26
< date: Mon, 10 Jun 2024 00:04:33 GMT
< vary: Origin
< x-robots-tag: noindex
< referrer-policy: no-referrer
< x-frame-options: DENY
< www-authenticate: Basic
< x-content-type-options: nosniff
< strict-transport-security: max-age=63072000
< server: am/2
< x-cache: Error from cloudfront
< via: 1.1 f3ac324bf05099849ebda59e8136db0e.cloudfront.net (CloudFront)
< x-amz-cf-pop: SEA19-C3
< x-amz-cf-id: km1QdjCrTYAjDuTOkwpGY6PUFs523KFX6djj0YZ-OPz4ROkjqyUKCA==
<
* Connection #0 to host secure.soundcloud.com left intact
{"error":"invalid_client"}%
```

**Comment by hubentom** [comment]
we are getting an invalid_grant error ... very inconsistently

**Comment by icreatestuff** [comment]
I'm also seeing this behaviour having updated to the new OAuth 2.1 method.

**Comment by hubentom** [comment]
is anyone from Soundcloud able to comment on this please?

**Comment by hubentom** [comment]
still no response? anyone else?

**Comment by MariaVWSoundCloud** [comment]
Hi all, let's first rule out, that your credentials are still valid. Could you please open a ticket with our support and mention "API access" so it will be routed quickly and correct?

**Comment by matthewmcvickar** [comment]
Thank you @MariaVWSoundCloud; I just filed a support request and put 'API Access' in the subject line. (I had already submitted one earlier this month as well.)

**Comment by matthewmcvickar** [comment]
@MariaVWSoundCloud I submitted a ticket almost two weeks ago and haven't heard any response. Can you check on the ticket or let me know when I might expect a reply?

**Comment by MariaVWSoundCloud** [comment]
@matthewmcvickar I can't find your ticket, unfortunately - can you post the ticket# here, please?

**Comment by matthewmcvickar** [comment]
@MariaVWSoundCloud Ticket number: 2980532

**Comment by MariaVWSoundCloud** [comment]
Ticket was resolved, I hope everything works now

**Comment by matthewmcvickar** [comment]
@MariaVWSoundCloud Huh? I received no reply to the ticket, and I still cannot authenticate with the API. Nothing has changed or been fixed.

**Comment by mgoodfellow** [comment]
Hey @matthewmcvickar 

I think the new auth docs have been confusing some people - my apologies if this isn't the mistake you are making but make sure you are doing it like this:

https://github.com/soundcloud/api/issues/288#issuecomment-2104854234

TL;DR;

Don't put the Base64() bit into the header, that's actually just telling you you need the output of a Base64 function and you replace the whole section after the Basic prefix.

**Comment by matthewmcvickar** [comment]
@mgoodfellow Ah, good to know. Unfortunately, I am still getting the same error when I remove the `Base64()`:

```
curl -v -X POST "https://secure.soundcloud.com/oauth/token" \
     -H  "accept: application/json; charset=utf-8" \
     -H  "Content-Type: application/x-www-form-urlencoded" \
     -H  "Authorization: Basic XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX:XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX" \
     --data-urlencode "grant_type=client_credentials"

Note: Unnecessary use of -X or --request, POST is already inferred.
* Host secure.soundcloud.com:443 was resolved.
* IPv6: (none)
* IPv4: 204.246.191.58, 204.246.191.54, 204.246.191.39, 204.246.191.110
*   Trying 204.246.191.58:443...
* Connected to secure.soundcloud.com (204.246.191.58) port 443
* ALPN: curl offers h2,http/1.1
* (304) (OUT), TLS handshake, Client hello (1):
*  CAfile: /etc/ssl/cert.pem
*  CApath: none
* (304) (IN), TLS handshake, Server hello (2):
* (304) (IN), TLS handshake, Unknown (8):
* (304) (IN), TLS handshake, Certificate (11):
* (304) (IN), TLS handshake, CERT verify (15):
* (304) (IN), TLS handshake, Finished (20):
* (304) (OUT), TLS handshake, Finished (20):
* SSL connection using TLSv1.3 / AEAD-AES128-GCM-SHA256 / [blank] / UNDEF
* ALPN: server accepted h2
* Server certificate:
*  subject: CN=*.soundcloud.com
*  start date: Feb  6 12:22:15 2024 GMT
*  expire date: Mar  9 12:22:14 2025 GMT
*  subjectAltName: host "secure.soundcloud.com" matched cert's "*.soundcloud.com"
*  issuer: C=BE; O=GlobalSign nv-sa; CN=GlobalSign GCC R3 DV TLS CA 2020
*  SSL certificate verify ok.
* using HTTP/2
* [HTTP/2] [1] OPENED stream for https://secure.soundcloud.com/oauth/token
* [HTTP/2] [1] [:method: POST]
* [HTTP/2] [1] [:scheme: https]
* [HTTP/2] [1] [:authority: secure.soundcloud.com]
* [HTTP/2] [1] [:path: /oauth/token]
* [HTTP/2] [1] [user-agent: curl/8.6.0]
* [HTTP/2] [1] [accept: application/json; charset=utf-8]
* [HTTP/2] [1] [content-type: application/x-www-form-urlencoded]
* [HTTP/2] [1] [authorization: Basic XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX:XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX]
* [HTTP/2] [1] [content-length: 29]
> POST /oauth/token HTTP/2
> Host: secure.soundcloud.com
> User-Agent: curl/8.6.0
> accept: application/json; charset=utf-8
> Content-Type: application/x-www-form-urlencoded
> Authorization: Basic XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX:XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX
> Content-Length: 29
>
< HTTP/2 401
< content-type: application/json; charset=utf-8
< content-length: 26
< date: Wed, 10 Jul 2024 18:20:23 GMT
< vary: Origin
< x-robots-tag: noindex
< referrer-policy: no-referrer
< x-frame-options: DENY
< www-authenticate: Basic
< x-content-type-options: nosniff
< strict-transport-security: max-age=63072000
< server: am/2
< x-cache: Error from cloudfront
< via: 1.1 c9cb74549b807d498500f04e1d4b8c7c.cloudfront.net (CloudFront)
< x-amz-cf-pop: HIO50-C2
< x-amz-cf-id: PhrKfLWr75oQD-fO419UdSCejPC4nIzx6aW2mrZ_q1omg9DVR8_H9g==
<
* Connection #0 to host secure.soundcloud.com left intact
{"error":"invalid_client"}%
```

**Comment by mgoodfellow** [comment]
You aren't converting it to base64 it looks like:

```
[HTTP/2] [1] [authorization: Basic XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX:XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX]
```

So before curl:

```
$ echo -n 'XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX:XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX' | base64
```

Take the output (which is the following for that bunch of X's):

`WFhYWFhYWFhYWFhYWFhYWFhYWFhYWFhYWFhYWFhYWFg6WFhYWFhYWFhYWFhYWFhYWFhYWFhYWFhYWFhYWFhYWFg=`

Now put that in the cURL command:

```
curl -v -X POST "https://secure.soundcloud.com/oauth/token" \
     -H  "accept: application/json; charset=utf-8" \
     -H  "Content-Type: application/x-www-form-urlencoded" \
     -H  "Authorization: Basic WFhYWFhYWFhYWFhYWFhYWFhYWFhYWFhYWFhYWFhYWFg6WFhYWFhYWFhYWFhYWFhYWFhYWFhYWFhYWFhYWFhYWFg=" \
     --data-urlencode "grant_type=client_credentials"
```

**Comment by mgoodfellow** [comment]
Oh never mind, you got there before me! Yeah, it caught me out first time around as well - glad you are sorted now!

**Comment by touchbee** [comment]
@mgoodfellow Hi, I'm currently converting my app as well to use the new Oauth Api. However, after I made the changes, I've noticed I'm getting the following error which makes it hard for me to test if it works. Also, I'm fairly certain I'm way under the rate limit defined here: https://developers.soundcloud.com/docs/api/rate-limits#play-requests. So not sure if I'm getting this error in "error" or if this is correct.

```
{
    "error": "rate_limit_exceeded"
}
```

**Comment by mgoodfellow** [comment]
@touchbee I'm not a SoundCloud employee, but I can probably point you in the right direction nonetheless :)

Are you using the Client Credentials flow by any chance? https://developers.soundcloud.com/docs/api/guide#client-creds

> Please be aware there is rate limiting on the amount of tokens you can request through the Client Credentials Flow: 50 tokens in 12h per app, and 30 tokens in 1h per IP address. In order to not hit the limit we highly recommend reusing one token between instances of your service and implementing the Refresh Token flow to renew tokens.

**Comment by youssefhassan** [comment]
@mgoodfellow thank you very much for being a top contributor here, much appreciated
@touchbee Sorry you are getting ratelimited while you shouldn't at least per your usage. Which endpoint are you getting ratelimited on and if you can share the soundcloud username or profile so I can dig deeper into the issue. Appreciate your feedback and thanks for your patience
